### DISCUSSION TITLE
Logging traffic summary after client disconnect

### DISCUSSION THREAD

**Discussion by lexandrop** [strongswan/strongswan#547] [Q&A] [answered]
Hi.

**_Question_**: is there any way to log traffic usage summary after connection is closed?

Currently the only place, where I can find traffic usage - is checking `ipsec statusall` while connection is still active, for example:
```
$ sudo ipsec statusall
Status of IKE charon daemon (strongSwan 5.8.2, Linux 5.4.0-73-generic, x86_64):
  uptime: 63 minutes, since Aug 09 16:00:26 2021
  malloc: sbrk 2277376, mmap 0, used 1631072, free 646304
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 14
  loaded plugins: charon aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink resolve socket-default connmark farp stroke updown eap-identity eap-aka eap-md5 eap-gtc eap-mschapv2 eap-dynamic eap-radius eap-tls eap-ttls eap-peap eap-tnc xauth-generic xauth-eap xauth-pam tnc-tnccs dhcp lookip error-notify certexpire led addrblock unity counters
Virtual IP pools (size/online/offline):
  10.10.10.0/24: 254/12/1
Listening IP addresses:
  ipX.X.X.X
Connections:
   ikev2-vpn:  %any...%any  IKEv2, dpddelay=300s
   ikev2-vpn:   local:  [ipX.X.X.X] uses public key authentication
   ikev2-vpn:    cert:  "CN=ipX.X.X.X"
   ikev2-vpn:   remote: uses EAP_RADIUS authentication with EAP identity '%any'
   ikev2-vpn:   child:  0.0.0.0/0 === dynamic TUNNEL, dpdaction=clear
Security Associations (12 up, 0 connecting):
   ikev2-vpn[61]: ESTABLISHED 65 seconds ago, ipX.X.X.X[ipX.X.X.X]...ipX.X.X.X[username1]
   ikev2-vpn[61]: IKEv2 SPIs: 0fdd8a93b3e872ee_i 03d566236e09eda6_r*, rekeying disabled
   ikev2-vpn[61]: IKE proposal: CHACHA20_POLY1305/PRF_HMAC_SHA2_512/CURVE_25519
   ikev2-vpn{16}:  INSTALLED, TUNNEL, reqid 16, ESP in UDP SPIs: cf27d425_i 865c42d1_o
   ikev2-vpn{16}:  CHACHA20_POLY1305, 129097 bytes_i (700 pkts, 0s ago), 365086 bytes_o (619 pkts, 0s ago), rekeying disabled
   ikev2-vpn{16}:   0.0.0.0/0 === 10.10.10.13/32
...
```
line
```
ikev2-vpn{16}:  CHACHA20_POLY1305, 129097 bytes_i (700 pkts, 0s ago), 365086 bytes_o (619 pkts, 0s ago), rekeying disabled
```

I've read the [introduction](https://wiki.strongswan.org/projects/strongswan/wiki/IntroductionTostrongSwan) and understand that
> The actual IPsec traffic is not handled by strongSwan but instead by the network and IPsec stack of the operating
> system kernel.

, but for some reason I assume (hope?), that charon is able to log this information. 

I've experimented with logger settings, checking info from different submodules on different logging levels, going from 1 to 4 - but did not find traffic summary being logged on any level.

Since my question is pretty general, I don't see great reasons to post my configs, log examples etc.

**Comment by tobiasbrunner** [comment, marked as answer]
> is there any way to log traffic usage summary after connection is closed?

That's already the case when a CHILD_SA is closed (not when it expires as it will already be gone already and no usage data can be retrieved from the kernel). You'll see something like: `[IKE] closing CHILD_SA ...{...} with SPIs ..._i (... bytes) ..._o (... bytes) and TS ... === ...`, no packet count is logged, though.

You might want to consider using [RADIUS accounting](https://wiki.strongswan.org/projects/strongswan/wiki/EapRadius#Accounting) to get accumulated traffic data (tracked over IKE and Child SA rekeyings during the same client session). It doesn't require RADIUS authentication to use it.

**Comment by lexandrop** [comment]
Hi @tobiasbrunner , thanks for your reply (especially for pointing out RADIUS accounting possibility) and sorry for the delay.
I'm marking your answer as best because I've answered exactly what I asked :)

> That's already the case when a CHILD_SA is closed (not when it expires as it will already be gone already and no usage data can be retrieved from the kernel). You'll see something like: [IKE] closing CHILD_SA ...{...} with SPIs ..._i (... bytes) ..._o (... bytes) and TS ... === ..., no packet count is logged, though.

For some reason, during my tests (I was using Android "strongSwan" app as a client) I never saw this line once, only saw it after you mentioned it and I checked server logs for some other clients.

> You might want to consider using RADIUS accounting to get accumulated traffic data (tracked over IKE and Child SA rekeyings during the same client session). It doesn't require RADIUS authentication to use it.

That's a useful feature, but, in my case, initially I was going to parse server's logs to get connections' info (including bandwidth) to something like logstash, so I can handle this data as I want further (processing/visualising).
If this data will be stored in RADIUS db, it adds some difficulties to my idea :) At this point I even consider constantly dumping `ipsec statusall` and getting information (about **current** connections ofc) from there.

To be honest, even ignoring the bandwidth usage info, I find strongSwan logs pretty hard to parse. I was making something similar for libreSwan/pluto+IKEv1 before (no holywars intended), and I could find many identifiers like ISAKMP/IPsec SA ids in every line, etc. I'm not sure, maybe it heavily dependps on IKEv1/IKEv2, but, so far, in strongSwan/charon+IKEv2 I see only something like `15[ENC]` or `14[IKE]` as a potential identifiers in every log line (and they repeat regularly), so at this point I'm not even sure how should I proceed with my idea.

Question: maybe there is some way to make strongSwan logs more "parsable"? For example, force-printing SA ids in every log line?
Any info (property name, documentation page, just an idea) would be useful.

**Reply by tobiasbrunner** [reply]
> Question: maybe there is some way to make strongSwan logs more "parsable"? For example, force-printing SA ids in every log line?
> Any info (property name, documentation page, just an idea) would be useful.

You might want to read [LoggerConfiguration](https://wiki.strongswan.org/projects/strongswan/wiki/LoggerConfiguration). But the log is really not intended to be machine-readable (in particular if you intend to track rekeyings etc.). Instead write a custom plugin or [vici event listener](https://wiki.strongswan.org/projects/strongswan/wiki/Vici) to collect information about SAs (or use the existing options like RADIUS accounting, which, by the way, does not necessarily require a RADIUS server, a simple Python daemon could probably store [the information in accounting messages](https://wiki.strongswan.org/projects/strongswan/wiki/EapRadius#Attributes-sent-to-RADIUS-servers)  anywhere you like).
